### PR TITLE
ClassDefinitionFixer - overhaul

### PR DESCRIPTION
**Pull request by SpacePossum**
Replaces: 
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2101
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1979

Closes:
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/2053
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1976 1 of 2

Addition fixes:
- (proper) support for anonymous classes
- (proper) support for more namespace madness

TODO
- [ ] master version
- [x] configuration options
- [x] check/test priority
- [x] check/test against other code bases (master SF and Twig tested, no hits)

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Why delete that integration test?

**Comment by SpacePossum** [comment]
The new fixer doesn't have the priority issue the current one has, so the test has become false.

**Comment by GrahamCampbell** [comment]
Would be good to keep the test though to prove it still works.

**Comment by SpacePossum** [comment]
we don't have a test to prove there is _not_ a priority issue, as the possible combinations with all the fixers against each would be extreme

**Comment by keradus** [comment]
Actually, we have:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/tree/1.12/Symfony/CS/Tests/Fixtures/Integration/misc

For sure it's not needed to have all combinations, but in rare cases when we want to check if bunch of fixers are working properly together, even without priority issue, we still could do that.

**Comment by SpacePossum** [comment]
For now I get;
```
1) Symfony\CS\Tests\IntegrationTest::testIntegration with data set #43 (Symfony\CS\Test\IntegrationCase Object (...))
Set priorities must be significant. If fixers used in reverse order return same output then the integration test is not sufficient or the priority relation between used fixers should not be set.
Failed asserting that two strings are not identical.
```

so that makes it legit to remove the priority test.
But sure I can make something else if it is needed.

**Comment by keradus** [comment]
I didn't go that deep, I am not saying is it needed in this case or not.
But it's possible, eg: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.12/Symfony/CS/Tests/Fixtures/Integration/misc/php_closing_tag%2Cshort_tag.test#L7

**Comment by SpacePossum** [comment]
ready for review

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php [hunk 1]
```diff
@@ -24,17 +25,61 @@
...
+     * @var array<string, bool>
+     */
+    private static $defaultConfig = array(
+        // put class declaration on one line
+        'singleLine' => false,
+        // if a classy extends or implements only one element than put it on the same line
+        'singleItemSingleLine' => false,
+        // if an interface extends multiple interfaces declared over multiple lines put each interface on its own line
+        'multiLineExtendsEachSingleLine' => false,
+    );
```
Comment (keradus) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: so, could I pass multiple true ?

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php [hunk 2]
```diff
@@ -24,17 +25,61 @@
...
+
+    /**
+     * @var array
+     */
+    private static $config;
```
Comment (keradus) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: what is default configuration?
what option is PSR-2 compatible ?

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php (hunk 1 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: depends on what you want, I struggled with the naming of the options so I'm open for other names :)

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php (hunk 2 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: current default is part of PSR 2 (spacing), but not all of it (i.e. make multi line etc.)

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php (hunk 1 above)
Comment (keradus) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: So I could use as config this array?
```php
 array(
        // put class declaration on one line
        'singleLine' => true,
        // if a classy extends or implements only one element than put it on the same line
        'singleItemSingleLine' => true,
        // if an interface extends multiple interfaces declared over multiple lines put each interface on its own line
        'multiLineExtendsEachSingleLine' => true,
    )
```

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php (hunk 2 above)
Comment (keradus) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: hard to see it, as default config is false everywhere... i got lost

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php (hunk 1 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: yes, that would work fine :)

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php (hunk 2 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: SF2:
```php
private static $defaultConfig = array(
    'singleLine' => true,
    'singleItemSingleLine' => false, // doesn't matter
    'multiLineExtendsEachSingleLine' => false, // doesn't matter
);
```

PSR2:
```php
private static $defaultConfig = array(
    'singleLine' => false,
    'singleItemSingleLine' => false,
    'multiLineExtendsEachSingleLine' => true,
);
```

Current default of the fixer:
```php
private static $defaultConfig = array(
    'singleLine' => false,
    'singleItemSingleLine' => true,
    'multiLineExtendsEachSingleLine' => true,
);
```

File: Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php (hunk 1 above)
Comment (keradus) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: how could I "put all on one line" and "put interfaces in their own lines" at the same time ?

Comment (SpacePossum) [inline Symfony/CS/Fixer/PSR2/ClassDefinitionFixer.php]: that config wouldn't make sense...
I can make some changes, not sure where to go though

File: Symfony/CS/Tests/Fixtures/Integration/priority/class_definition,trailing_spaces.test [hunk 3]
```diff
@@ -1,22 +0,0 @@
...
-{
-}
-
---INPUT--
-<?php
-class Aaa implements
-    Symfony\CS\Tests\Fixer,\RFb,
-    \Fcc1, \GFdd
-{
-}
```
Comment (keradus) [inline Symfony/CS/Tests/Fixtures/Integration/priority/class_definition,trailing_spaces.test]: could we recover that test back and put it in `misc` and test it without priority check ?

**Comment by keradus** [comment]
Please validate the input to configuration.
Having all false is pointless, having conflicting settings is bad as well.

BTW, the RuleSet will need to be updated on 2.x line. Also, default setting on 2.x should be pure PSR2 as usuall.

**Comment by SpacePossum** [comment]
>Having all false is pointless

It is not, the fixer will still fix the spacing between keywords following PSR2. It will however not fix multiline/single line, it will not touch that part of classy definitions.

> having conflicting settings is bad [sic].

Good point, I'll update the configuration check.

>BTW, the RuleSet will need to be updated on 2.x line. Also, default setting on 2.x should be pure PSR2 as usuall.

Yes, making a master version of this has been on the TODO list of the PR since it was opened, I'll follow up on that as soon as 1.x has been done :)

**Comment by keradus** [comment]
:+1:

**Comment by keradus** [comment]
@SpacePossum is something still needed here? Is this PR a WIP or not ?

**Comment by SpacePossum** [comment]
IMHO this is RTM :)

**Comment by GrahamCampbell** [comment]
I thought PSR-2 was a strict subset of Symfony, so how come symfony have not followed psr-2 in that one case we have here?

**Comment by keradus** [comment]
PSR2: **if** you would write it in many lines, then ...
Symfony: don't write it in many lines

**Comment by SpacePossum** [comment]
PSR 2 
>Lists of implements MAY be split across multiple line

SF doesn't want that, so they do follow PSR2, but you can configure the fixer to follow PSR2 in another way

**Comment by GrahamCampbell** [comment]
Ah, I see. :+1:

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by keradus** [comment]
Awesome work @SpacePossum !
just got merged into 2.x line, please prepare PR for rulesets ;)

**Pull request closed without merging**